### PR TITLE
TArrayDeque fix for .remove(Object)

### PR DESCRIPTION
**Pull request by adamjryan**
ArrayDeque removes items by shifting the existing items to overwrite the removed item. It did not update the head/tail pointers once the operation was complete leaving a null item at the head/tail of the collection. This change updates the pointers so that the null element is excluded and the correct size is returned.

### CODE REVIEW

**Comment by konsoletyper** [comment]
Did you see my code review comments?

**Comment by adamjryan** [comment]
If you mean on this branch, then no I can't see them.

File: classlib/src/main/java/org/teavm/classlib/java/util/TArrayDeque.java [hunk 1]
```diff
@@ -254,24 +254,28 @@ private void removeAt(int index) {
...
                 }
-                array[tail - 1] = null;
+                tail -= 1;
+                array[tail] = null;
             } else {
                 for (int i = index - 1; i >= head; --i) {
                     array[i + 1] = array[i];
                 }
                 array[head] = null;
+                head++;
```
[the remark this answers is not in the document]
Comment (adamjryan) [reply classlib/src/main/java/org/teavm/classlib/java/util/TArrayDeque.java]: Heh, that was the first thing I tried but it failed the build due to some code quality tool. This is the same reason for the changes below as well.

[the remark this answers is not in the document]
Comment (konsoletyper) [reply classlib/src/main/java/org/teavm/classlib/java/util/TArrayDeque.java]: TeaVM uses checkstyle to verify code style. I set up rules by my own, and, yes, I prohibited `=` expressions, so you can't write something like `array[head += 1] = null`. But using `++` and `--` within expressions is ok.

File: classlib/src/test/java/org/teavm/classlib/java/util/TestTArrayDeque.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package org.teavm.classlib.java.util;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+public class TestTArrayDeque {
```
[the remark this answers is not in the document]
Comment (adamjryan) [reply classlib/src/test/java/org/teavm/classlib/java/util/TestTArrayDeque.java]: Using Test as a prefix stops test classes from polluting the auto-complete when working with the domain types. It also makes it easier to read the test class names in the project view listings when they all have a prefix of the same length.

[the remark this answers is not in the document]
Comment (konsoletyper) [reply classlib/src/test/java/org/teavm/classlib/java/util/TestTArrayDeque.java]: Ok, you may be right, but you see, you are contributing to a large project with existing rules and code conventions. You should follow these, not your own.

**Comment by konsoletyper** [comment]
Thanks! I know, this was quite challenging!

**Comment by adamjryan** [comment]
True, for a four liner. Some of my changes might have gone against the grain, it's only to offer alternate approaches and maybe not the final state. Btw to keep the history clean I was going to squash the commit before the merge, means the changes are more atomic when looking back.

**Comment by adamjryan** [comment]
Ah, I see you already did that. Just realised there's module "tests" in the project for the ClassLib tests that these ones should have gone in.

**Comment by konsoletyper** [comment]
I see. Also, they are marked with custom runner (which transform these tests into JavaScripts and runs via HtmlUnit). I'll move the test myself

**Pull request merged**